### PR TITLE
to_marc21: generate xml

### PR DESCRIPTION
**Pull request by greut**
* Work in progress MARC XML serialization.

**TODO**
* [ ] serialize to JSON
* [ ] immutable data structure

Signed-off-by: Yoan Blanc <yoan@dosimple.ch>

### CODE REVIEW

File: dojson/contrib/to_marc21/utils.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+from __future__ import unicode_literals
+
+import pkg_resources
+
+from lxml import etree
+from lxml.builder import E
+
+
+MARC21_DTD = pkg_resources.resource_filename(
```
Comment (jirikuncar) [inline dojson/contrib/to_marc21/utils.py]: shall we reuse the one from `dojson.contrib.marc21.utils`?

File: dojson/contrib/marc21/utils.py [hunk 2]
```diff
@@ -139,3 +122,77 @@ def load(source):
...
         yield create_record(data)
+
+
+class GroupableOrderedDict(MutableMapping):
+    """List that can group values pretending to be a dict."""
+
+    def __init__(self, values=None):
+        super(MutableMapping, self).__init__()
+
+        self._data = []
```
Comment (jirikuncar) [inline dojson/contrib/marc21/utils.py]:
```python
self._data = defaultdict(list)
self._keys = []
for k, v in values or []:
    self._data[k].append(v)
    self._keys.append(v)
```

Comment (PXke) [inline dojson/contrib/marc21/utils.py]: Theorical question, but why don't we put values to () as default value ?

File: dojson/contrib/marc21/utils.py [hunk 3]
```diff
@@ -139,3 +122,77 @@ def load(source):
...
+        self._keys = []
+        if values:
+            for k, v in values:
+                self[k] = v
+
+    def get(self, key):
+        if key in self._keys:
```
Comment (PXke) [inline dojson/contrib/marc21/utils.py]: Sorry but you are calling __getitem__ which is also doing the check of 
```
if key in self._keys:
```

It seems a 
```
try:
    do_my_things()
except KeyError:
    return None
```
would be more logic what do you think ? 

Also could get take a default value ?  :)

File: dojson/contrib/marc21/utils.py [hunk 4]
```diff
@@ -139,3 +122,77 @@ def load(source):
...
+            return self.__getitem__(key)
+        else:
+            return None
+
+    def __getitem__(self, key):
+        if key in self._keys:
+            output = [v for k, v in self._data if k == key]
+            if len(output) == 1:
+                return output[0]
```
Comment (PXke) [inline dojson/contrib/marc21/utils.py]: I know python is a lot of ducktyping, but in this case this function will return two differents type. Which can be a bit annoying what do you think ? Could we always return a list ?

File: dojson/contrib/marc21/utils.py [hunk 5]
```diff
@@ -139,3 +122,77 @@ def load(source):
...
+
+    def __delitem__(self, key):
+        self._data = [(k, v) for k, v in self._data if k != key]
+        self._keys.remove(key)
+
+    def __len__(self):
+        return len(self.keys)
+
+    def values(self, **kw):
+        r = []
```
Comment (PXke) [inline dojson/contrib/marc21/utils.py]: Sorry but could you add a little docstring with the kw value and what it does :) ? I know we can read the code, but **kwargs are always a bit nebulous for new developper.

File: dojson/contrib/marc21/utils.py [hunk 6]
```diff
@@ -139,3 +122,77 @@ def load(source):
...
+        if not kw.get('expand', False):
+            for key in self._keys:
+                r.append(self[key])
+        else:
+            for _, v in self._data:
+                r.append(v)
+        return r
+
+    def keys(self, **kw):
+        if not kw.get('repeated', False):
```
Comment (PXke) [inline dojson/contrib/marc21/utils.py]: Cool option !

File: dojson/contrib/to_marc21/utils.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+def dumps(*records, **kwargs):
+    """Dump records into a MarcXML file."""
+
+    root = etree.Element('collection', nsmap={None: MARC21_NS})
+    for record in records:
+        rec = E.record()
+        for df, subfields in record.items():
+            if isinstance(subfields, list):
+                controlfield = E.controlfield(subfields[0])
+                print(df)
```
Comment (PXke) [inline dojson/contrib/to_marc21/utils.py]: missing debugging instruction ?

File: dojson/contrib/marc21/utils.py (hunk 3 above)
Comment (greut) [inline dojson/contrib/marc21/utils.py]: yup!

File: dojson/contrib/marc21/utils.py (hunk 4 above)
Comment (greut) [inline dojson/contrib/marc21/utils.py]: yup

File: dojson/contrib/to_marc21/utils.py (hunk 7 above)
Comment (greut) [inline dojson/contrib/to_marc21/utils.py]: indeed.

**Pull request closed without merging**